
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://chemkg.github.io/chemrof/">
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.3, mkdocs-material-8.1.8">
    
    
      
        <title>CHEMROF</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.6e60f8b8.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.e6a45f82.min.css">
        
          
          
          <meta name="theme-color" content="#ab47bd">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="purple" data-md-color-accent="">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#chemrof-schema" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="CHEMROF" class="md-header__button md-logo" aria-label="CHEMROF" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            CHEMROF
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Schema
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/chemkg/chemrof/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="CHEMROF" class="md-nav__button md-logo" aria-label="CHEMROF" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    CHEMROF
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/chemkg/chemrof/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="home/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="database/" class="md-nav__link">
        Database
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="ontology/" class="md-nav__link">
        Ontology
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Schema
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="." class="md-nav__link md-nav__link--active">
        Schema
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#classes" class="md-nav__link">
    Classes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#mixins" class="md-nav__link">
    Mixins
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#slots" class="md-nav__link">
    Slots
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#enums" class="md-nav__link">
    Enums
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subsets" class="md-nav__link">
    Subsets
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#types" class="md-nav__link">
    Types
  </a>
  
    <nav class="md-nav" aria-label="Types">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#built-in" class="md-nav__link">
    Built in
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#defined" class="md-nav__link">
    Defined
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="metaclasses/" class="md-nav__link">
        Metaclasses
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="credits/" class="md-nav__link">
        Credits
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#classes" class="md-nav__link">
    Classes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#mixins" class="md-nav__link">
    Mixins
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#slots" class="md-nav__link">
    Slots
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#enums" class="md-nav__link">
    Enums
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subsets" class="md-nav__link">
    Subsets
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#types" class="md-nav__link">
    Types
  </a>
  
    <nav class="md-nav" aria-label="Types">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#built-in" class="md-nav__link">
    Built in
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#defined" class="md-nav__link">
    Defined
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
<a href="https://github.com/chemkg/chemrof/edit/master/docs/index.md" title="Edit this page" class="md-content__button md-icon">
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
</a>


<h1 id="chemrof-schema">chemrof schema</h1>
<p>A data model for managing information about chemical entities, ranging from atoms
through molecules to complex mixtures.</p>
<p>Aspects of this have been cribbed from various sources including CHEBI, SIO,
Wikipedia/Wikidata, the NCATS Translator Chemical Working Group, but all mistakes
are my own.</p>
<p>For full context/motivation see the <a href="https://github.com/chemkg/chemrof">GitHub repo</a>.</p>
<h3 id="classes">Classes</h3>
<ul>
<li><a href="ChemicalEntity/">ChemicalEntity</a> - An entity that can be described using using chemical properties.<ul>
<li><a href="Atom/">Atom</a> - A chemical entity consisting of exactly one atomic nucleus and the electron(s) orbiting it.<ul>
<li><a href="FullySpecifiedAtom/">FullySpecifiedAtom</a> - An atom type that has atomic number, charge, and neutron number stated (equivalently: protons + mass + charge)</li>
<li><a href="PartiallySpecifiedAtom/">PartiallySpecifiedAtom</a> - An atom type that only has a subset of properties (has atomic number, charge, and neutron number) stated<ul>
<li><a href="AtomIonicForm/">AtomIonicForm</a> - an atom type in which only charge state is specified, designated by the element followed by charge<ul>
<li><a href="MonoatomicIon/">MonoatomicIon</a> - An atom type that has a charge<ul>
<li><a href="AtomAnion/">AtomAnion</a> - A monoatomic ion that is negative</li>
<li><a href="AtomCation/">AtomCation</a> - A monoatomic ion that is positive, e.g Cu(2+)</li>
</ul>
</li>
<li><a href="UnchargedAtom/">UnchargedAtom</a> - An atom type that has no charge</li>
</ul>
</li>
<li><a href="ChemicalElement/">ChemicalElement</a> - generic form of an atom, with unspecified neutron or charge</li>
<li><a href="Nuclide/">Nuclide</a> - an atomic species characterized by the specific constitution of its nucleus<ul>
<li><a href="Isobar/">Isobar</a> - Isobars are atoms (nuclides) of different chemical elements that have the same number of nucleons.</li>
<li><a href="Isotope/">Isotope</a> - A specific subtype of an atom with a specified neutron number</li>
<li><a href="Radionuclide/">Radionuclide</a> - an atom that has excess nuclear energy, making it unstable</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><a href="PolyatomicEntity/">PolyatomicEntity</a> - Any chemical entity consisting of more than one atom.<ul>
<li><a href="ChemicalMixture/">ChemicalMixture</a> - A polyatomic entity that consists of two or more parts that are weakly bound, may be of known or unknown stoichiometry<ul>
<li><a href="ImpreciseChemicalMixture/">ImpreciseChemicalMixture</a> - A macroscopic polyatomic entity that consists of multiple chemical entities where the stoichiometry is not specified and the spatial arrangement of the ingredients is not defined. Examples include coal, tea tree oil, toothpaste, organometallic compounds, heterogeneous metals</li>
<li><a href="PreciseChemicalMixture/">PreciseChemicalMixture</a> - A polyatomic entity that consists of two or more parts that are weakly bound, and of known stoichiometry<ul>
<li><a href="ChemicalSalt/">ChemicalSalt</a> - a chemical compound consisting of an ionic assembly of cations and anions.</li>
<li><a href="MolecularComplex/">MolecularComplex</a> - A molecular complex is a chemical complex composed of at least one weakly interacting molecule.[SIO] Examples: chromosome, double-standed DNA, protein-containing macromolecular complex<ul>
<li><a href="SupramolecularPolymer/">SupramolecularPolymer</a> - polymeric arrays of monomeric units that are connected by reversible and highly directional secondary interactions–that is, non-covalent bonds</li>
</ul>
</li>
<li><a href="RacemicMixture/">RacemicMixture</a> - a chemical compound that has equal amounts of left- and right-handed enantiomers of a chiral molecule. An example is Thalidomide</li>
</ul>
</li>
</ul>
</li>
<li><a href="MolecularComponent/">MolecularComponent</a> - A proper part of a molecule that is larger than an atom<ul>
<li><a href="ChemicalGroup/">ChemicalGroup</a></li>
<li><a href="FunctionalGroup/">FunctionalGroup</a></li>
<li><a href="Moiety/">Moiety</a> - a named part of a molecule. In some instances moieties may be composed of yet smaller moieties and functional groups</li>
<li><a href="PolymerPart/">PolymerPart</a><ul>
<li><a href="MolecularSubsequence/">MolecularSubsequence</a></li>
<li><a href="Monomer/">Monomer</a></li>
</ul>
</li>
<li><a href="SequenceInterval/">SequenceInterval</a> - a chain of connected monomers in a linear polymer - for example, a sub-region of an amino acid or nucleotide sequence.<ul>
<li><a href="AminoAcidSequenceInterval/">AminoAcidSequenceInterval</a></li>
<li><a href="NucleotideSequenceInterval/">NucleotideSequenceInterval</a><ul>
<li><a href="DNASequenceInterval/">DNASequenceInterval</a></li>
<li><a href="RNASequenceInterval/">RNASequenceInterval</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><a href="Molecule/">Molecule</a> - A chemical entity that consists of two or more atoms where all atoms are connected via covalent bonds.<ul>
<li><a href="BronstedAcid/">BronstedAcid</a> - A molecule capable of donating a hydron to an acceptor. Example: citric acid</li>
<li><a href="Allotrope/">Allotrope</a> - A molecule consisting of atoms of the same element</li>
<li><a href="Ester/">Ester</a> - An ester is a chemical compound derived from an acid (organic or inorganic) in which at least one –OH hydroxyl group is replaced by an –O– alkyl (alkoxy) group</li>
<li><a href="Macromolecule/">Macromolecule</a> - A large molecule, such as a polymer, lipid, or macrocycle<ul>
<li><a href="Glycan/">Glycan</a> - A macromolecule consisting of a large number of monosaccharides linked glycosidically</li>
<li><a href="Lipid/">Lipid</a></li>
<li><a href="MonomolecularPolymer/">MonomolecularPolymer</a> - A macromolecule composed of repeating subunits. Examples: cellulose<ul>
<li><a href="Copolymer/">Copolymer</a> - a polymer derived from more than one species of monomer</li>
</ul>
</li>
<li><a href="Peptide/">Peptide</a></li>
<li><a href="Protein/">Protein</a></li>
</ul>
</li>
<li><a href="MolecularSpecies/">MolecularSpecies</a> - A molecule in which the units are identical. Example: methyl</li>
<li><a href="NaturalProduct/">NaturalProduct</a></li>
<li><a href="NonSpeciesMolecule/">NonSpeciesMolecule</a> - A molecule in which the units are not identical. TODO: move this higher</li>
<li><a href="PolyatomicIon/">PolyatomicIon</a> - A molecule that has a charge. For example, nitrate is a molecular and ionic species; its formula is (NO3-).<ul>
<li><a href="MolecularAnion/">MolecularAnion</a> - A polyatomic ion that is negative<ul>
<li><a href="AcidBase/">AcidBase</a> - A conugate base of an acid with charge specified. Example: citrate(3-)</li>
</ul>
</li>
<li><a href="MolecularCation/">MolecularCation</a> - A polyatomic ion that is positive</li>
</ul>
</li>
<li><a href="SmallMolecule/">SmallMolecule</a> - A molecule characterized by availability in small-molecule databases</li>
<li><a href="Stereoisomer/">Stereoisomer</a><ul>
<li><a href="Enantiomer/">Enantiomer</a> - one of two stereoisomers of a chiral molecule that are mirror images. Example: R-thalidomide</li>
</ul>
</li>
<li><a href="UnchargedMolecule/">UnchargedMolecule</a> - A molecule that has no charge</li>
</ul>
</li>
</ul>
</li>
<li><a href="SubatomicParticle/">SubatomicParticle</a> - A chemical entity below the granularity of an atom.<ul>
<li><a href="Electron/">Electron</a></li>
<li><a href="Nucleon/">Nucleon</a> - A neutron or proton<ul>
<li><a href="Neutron/">Neutron</a></li>
<li><a href="Proton/">Proton</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><a href="ChemicalRole/">ChemicalRole</a> - A characteristic of a chemical entity that is realized under particular conditions</li>
<li><a href="GroupingClass/">GroupingClass</a> - A non-specific entity. For example "amino acid" is a grouping for "serine", "leucine" etc; ester is a grouping for nitroglycerin; alkane is a grouping for pentane, butane, methane, etc<ul>
<li><a href="AtomGroupingClass/">AtomGroupingClass</a> - A grouping class that classifies atoms. For example, metal atom<ul>
<li><a href="AtomGroupingByPeriodicTablePlacement/">AtomGroupingByPeriodicTablePlacement</a><ul>
<li><a href="AtomGroupingByPeriodicTableBlock/">AtomGroupingByPeriodicTableBlock</a> - Example: group 13 atom</li>
<li><a href="AtomGroupingByPeriodicTableGroup/">AtomGroupingByPeriodicTableGroup</a> - Example: p-block atom</li>
</ul>
</li>
<li><a href="AtomGroupingByProperty/">AtomGroupingByProperty</a> - Example: metal atom</li>
</ul>
</li>
<li><a href="ChemicalGroupingByCharge/">ChemicalGroupingByCharge</a> - A grouping class that refines a charge-neutral entity</li>
<li><a href="MolecularComponentGroupingClass/">MolecularComponentGroupingClass</a> - A grouping class that classifies molecular components. Example: inorganic anion group</li>
<li><a href="MoleculeGroupingClass/">MoleculeGroupingClass</a> - A grouping class that classifies molecules. Example: carbohydrate, monosaccharide, amino acid, polyacrylamide<ul>
<li><a href="ChemicalSaltGroupingClass/">ChemicalSaltGroupingClass</a> - A grouping class for a chemical salt<ul>
<li><a href="ChemicalSaltByAnion/">ChemicalSaltByAnion</a> - A grouping class for a chemical salt named by anion</li>
<li><a href="ChemicalSaltByCation/">ChemicalSaltByCation</a> - A grouping class for a chemical salt named by cation</li>
</ul>
</li>
<li><a href="GeneralizedMolecularStructureClass/">GeneralizedMolecularStructureClass</a> - A molecule grouping class that can be written using a chemical formula using variables (e.g. n), or arithmetic expressions uing variables (e.g. 2n+2), plus optionally constraints on those variables (e.g. n&gt;3). An example is alkane; CnH2n+2, which groups specific molecules such as methane</li>
<li><a href="GroupingClassForAcidsOrBases/">GroupingClassForAcidsOrBases</a><ul>
<li><a href="AcidAnionGroupingClass/">AcidAnionGroupingClass</a> - A molecule grouping class that groups the different ion forms of the base of an acid</li>
<li><a href="AcidBaseConflationClass/">AcidBaseConflationClass</a> - A molecule grouping class that can groups an acid together with all its conjugate bases. physiological</li>
<li><a href="GeneralAcidBaseGroupingClass/">GeneralAcidBaseGroupingClass</a> - A molecule grouping class that can groups an acid together with all its conjugate bases. physiological</li>
</ul>
</li>
<li><a href="MolecularDerivativeGroupingClass/">MolecularDerivativeGroupingClass</a> - A grouping class defined as a derivative of another molecule or molecular grouping class</li>
<li><a href="MoleculeGroupingClassDefinedByAdditionOfAGroup/">MoleculeGroupingClassDefinedByAdditionOfAGroup</a> - A grouping class defined in terms of a derivative of a molecule and the addition of a group</li>
<li><a href="MoleculeGroupingClassDefinedByComponents/">MoleculeGroupingClassDefinedByComponents</a> - A grouping class defined as an exhaustive list of components</li>
</ul>
</li>
</ul>
</li>
<li><a href="InformationalChemicalEntity/">InformationalChemicalEntity</a><ul>
<li><a href="InchiObject/">InchiObject</a> - A representation of an InChI string as an object, composed of distinct sublayers<ul>
<li><a href="StandardInchiObject/">StandardInchiObject</a></li>
</ul>
</li>
<li><a href="PeriodicTablePlacement/">PeriodicTablePlacement</a><ul>
<li><a href="PeriodicTableBlock/">PeriodicTableBlock</a> - A block of the periodic table is a set of elements unified by the orbitals their valence electrons or vacancies lie in</li>
</ul>
</li>
</ul>
</li>
<li><a href="Location/">Location</a><ul>
<li><a href="Stereocenter/">Stereocenter</a> - a particular instance of a stereogenic element that is geometrically a point, bearing different substituents, such that interchanging any two substituents leads to a stereoisomer<ul>
<li><a href="ChiralityCenter/">ChiralityCenter</a> - a stereocenter consisting of an atom holding a set of ligands (atoms or groups of atoms) in a spatial arrangement which is not superimposable on its mirror image. </li>
</ul>
</li>
</ul>
</li>
<li><a href="Material/">Material</a> - A substance composed of one or more chemical entities<ul>
<li><a href="Nanostructure/">Nanostructure</a> - A nanostructure is a structure of intermediate size between microscopic and molecular structures. Nanostructural detail is microstructure at nanoscale.</li>
</ul>
</li>
<li><a href="Reaction/">Reaction</a> - reaction<ul>
<li><a href="IsomeraseReaction/">IsomeraseReaction</a> - A reaction that converts a molecule from one isomer to another.</li>
</ul>
</li>
<li><a href="RelationalChemicalEntity/">RelationalChemicalEntity</a> - A reified relationship<ul>
<li><a href="AtomOccurrence/">AtomOccurrence</a> - An occurrence of an atom in the context of a particular molecule. For example, one of two occurrences of oxygen in nitrate</li>
<li><a href="AtomicBond/">AtomicBond</a> - A connection between two atoms. Note this is the reified form of 'atomically connected to'. </li>
<li><a href="ProportionalPart/">ProportionalPart</a> - A part of a complex mixture that is of uniform composition</li>
<li><a href="ReactionParticipant/">ReactionParticipant</a> - Instances of this link a reaction to a chemical entity participant</li>
<li><a href="Similarity/">Similarity</a> - A relationship between two or more entities that is quantified based on their similarity<ul>
<li><a href="MoleculePairwiseSimilarity/">MoleculePairwiseSimilarity</a><ul>
<li><a href="TanimotoSimilarity/">TanimotoSimilarity</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<h3 id="mixins">Mixins</h3>
<ul>
<li><a href="AnionState/">AnionState</a> - This mixin is applied on an entity (atom or molecule) that is negatively charged</li>
<li><a href="CationState/">CationState</a> - This mixin is applied on an entity (atom or molecule) that is positively charged</li>
<li><a href="ChargeState/">ChargeState</a> - Applies to entities that have a charge state, which may be positive, negative, or neutral<ul>
<li><a href="Charged/">Charged</a><ul>
<li><a href="AnionState/">AnionState</a> - This mixin is applied on an entity (atom or molecule) that is negatively charged</li>
<li><a href="CationState/">CationState</a> - This mixin is applied on an entity (atom or molecule) that is positively charged</li>
</ul>
</li>
<li><a href="Uncharged/">Uncharged</a></li>
</ul>
</li>
<li><a href="Charged/">Charged</a><ul>
<li><a href="AnionState/">AnionState</a> - This mixin is applied on an entity (atom or molecule) that is negatively charged</li>
<li><a href="CationState/">CationState</a> - This mixin is applied on an entity (atom or molecule) that is positively charged</li>
</ul>
</li>
<li><a href="ChemicalEntityOrGrouping/">ChemicalEntityOrGrouping</a></li>
<li><a href="Connectivity/">Connectivity</a><ul>
<li><a href="FullyCovalentlyBonded/">FullyCovalentlyBonded</a> - All pairs of atoms in this entity are transitively connected via covalent bonds</li>
<li><a href="WeaklyBonded/">WeaklyBonded</a> - Not covalently bonded</li>
</ul>
</li>
<li><a href="EntityWithAtomsEnumerated/">EntityWithAtomsEnumerated</a> - A mixin for an entity that consists of one or more atoms where the atoms and the number of occurrences is counted (corresponding to the chemical layer in InChI)</li>
<li><a href="EntityWithChargeSpecified/">EntityWithChargeSpecified</a> - A mixin for an entity where the charge is explicitly specified (corresponding to the /q layer in InChI)</li>
<li><a href="EntityWithConnectivitySpecificied/">EntityWithConnectivitySpecificied</a> - A mixin for an entity that consists of two or more atom occurrences where the connectivity is specified (corresponding to the /c layer in InChI)</li>
<li><a href="EntityWithIsotopeSpecified/">EntityWithIsotopeSpecified</a> - A mixin for an entity where the charge is explicitly specified (corresponding to the /q layer in InChI)</li>
<li><a href="EntityWithStereochemistrySpecified/">EntityWithStereochemistrySpecified</a> - A mixin for an entity where the stereochemistry is explicitly specified (corresponding to the /t, m/, and /s layers in InChI)</li>
<li><a href="IsPartOrWhole/">IsPartOrWhole</a><ul>
<li><a href="IsPart/">IsPart</a></li>
<li><a href="IsWhole/">IsWhole</a></li>
</ul>
</li>
<li><a href="OwlClass/">OwlClass</a> - instances of this class mixin are OWL Classes</li>
<li><a href="SpecificityMixin/">SpecificityMixin</a><ul>
<li><a href="EntityWithAtomsEnumerated/">EntityWithAtomsEnumerated</a> - A mixin for an entity that consists of one or more atoms where the atoms and the number of occurrences is counted (corresponding to the chemical layer in InChI)</li>
<li><a href="EntityWithChargeSpecified/">EntityWithChargeSpecified</a> - A mixin for an entity where the charge is explicitly specified (corresponding to the /q layer in InChI)</li>
<li><a href="EntityWithConnectivitySpecificied/">EntityWithConnectivitySpecificied</a> - A mixin for an entity that consists of two or more atom occurrences where the connectivity is specified (corresponding to the /c layer in InChI)</li>
<li><a href="EntityWithIsotopeSpecified/">EntityWithIsotopeSpecified</a> - A mixin for an entity where the charge is explicitly specified (corresponding to the /q layer in InChI)</li>
<li><a href="EntityWithStereochemistrySpecified/">EntityWithStereochemistrySpecified</a> - A mixin for an entity where the stereochemistry is explicitly specified (corresponding to the /t, m/, and /s layers in InChI)</li>
</ul>
</li>
<li><a href="State/">State</a><ul>
<li><a href="ChargeState/">ChargeState</a> - Applies to entities that have a charge state, which may be positive, negative, or neutral<ul>
<li><a href="Charged/">Charged</a><ul>
<li><a href="AnionState/">AnionState</a> - This mixin is applied on an entity (atom or molecule) that is negatively charged</li>
<li><a href="CationState/">CationState</a> - This mixin is applied on an entity (atom or molecule) that is positively charged</li>
</ul>
</li>
<li><a href="Uncharged/">Uncharged</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="Uncharged/">Uncharged</a></li>
</ul>
<h3 id="slots">Slots</h3>
<ul>
<li><a href="absolute_configuration/">absolute configuration</a><ul>
<li><a href="enantiomer_absolute_configuration/">enantiomer➞absolute configuration</a></li>
</ul>
</li>
<li><a href="acidity/">acidity</a><ul>
<li><a href="Bronsted_acid_acidity/">Bronsted acid➞acidity</a></li>
</ul>
</li>
<li><a href="atomically_connected_to/">atomically connected to</a></li>
<li><a href="boiling_point_in_celcius/">boiling point in celcius</a><ul>
<li><a href="chemical_element_boiling_point_in_celcius/">chemical element➞boiling point in celcius</a></li>
</ul>
</li>
<li><a href="bond_angle/">bond angle</a><ul>
<li><a href="atomic_bond_bond_angle/">atomic bond➞bond angle</a></li>
</ul>
</li>
<li><a href="bond_energy/">bond energy</a> - the measure of bond strength in a chemical bond.</li>
<li><a href="bond_length/">bond length</a> - the average distance between nuclei of two bonded atoms in a molecule</li>
<li><a href="bond_length_in_angstroms/">bond length in angstroms</a> - the average distance between nuclei of two bonded atoms in a molecule. It is a transferable property of a bond between atoms of fixed types, relatively independent of the rest of the molecule.<ul>
<li><a href="atomic_bond_bond_length_in_angstroms/">atomic bond➞bond length in angstroms</a></li>
</ul>
</li>
<li><a href="bond_order/">bond order</a> - the number of electron pairs (bonds) between a pair of atoms</li>
<li><a href="bond_type/">bond type</a><ul>
<li><a href="atomic_bond_bond_type/">atomic bond➞bond type</a></li>
</ul>
</li>
<li><a href="charge_agnostic_entity/">charge agnostic entity</a><ul>
<li><a href="chemical_grouping_by_charge_charge_agnostic_entity/">chemical grouping by charge➞charge agnostic entity</a></li>
</ul>
</li>
<li><a href="chemical_property/">chemical property</a><ul>
<li><a href="decay_energy/">decay energy</a></li>
<li><a href="decay_mode/">decay mode</a></li>
<li><a href="decay_product/">decay product</a></li>
<li><a href="half_life/">half life</a></li>
<li><a href="has_cyclic_structure/">has cyclic structure</a> - true if the molecule possesses a cycle in the atom connectivity graph</li>
<li><a href="mode_of_formation/">mode of formation</a></li>
<li><a href="property_of_atom/">property of atom</a><ul>
<li><a href="elemental_charge/">elemental charge</a> - number of protons minus number of electrons<ul>
<li><a href="anion_state_elemental_charge/">anion state➞elemental charge</a><ul>
<li><a href="atom_anion_elemental_charge/">atom anion➞elemental charge</a></li>
</ul>
</li>
<li><a href="cation_state_elemental_charge/">cation state➞elemental charge</a><ul>
<li><a href="atom_cation_elemental_charge/">atom cation➞elemental charge</a></li>
</ul>
</li>
<li><a href="chemical_salt_elemental_charge/">chemical salt➞elemental charge</a></li>
<li><a href="uncharged_elemental_charge/">uncharged➞elemental charge</a></li>
</ul>
</li>
<li><a href="subatomic_particle_count/">subatomic particle count</a><ul>
<li><a href="atomic_number/">atomic number</a> - number of protons in an atom</li>
<li><a href="neutron_number/">neutron number</a> - number of neutrons in an atom</li>
<li><a href="nucleon_number/">nucleon number</a> - number of nucleons in an atom. proton number plus neutron number.</li>
</ul>
</li>
</ul>
</li>
<li><a href="thermophysical_property/">thermophysical property</a><ul>
<li><a href="molar_cohesive_energy/">molar cohesive energy</a></li>
<li><a href="molar_value/">molar value</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="chemical_to_chemical_relationship/">chemical to chemical relationship</a><ul>
<li><a href="alternate_form_of/">alternate form of</a> - A relationship between two entities of the same type<ul>
<li><a href="allotropic_analog_of/">allotropic analog of</a> - relationship that holds between two allotropes, where both allotropes are allotropes of the same element<ul>
<li><a href="allotrope_allotropic_analog_of/">allotrope➞allotropic analog of</a></li>
</ul>
</li>
<li><a href="chemical_isomer_of/">chemical isomer of</a> - a relationship that holds between two forms of a parent molecule with identical chemical formulas but distinct arrangements of atoms in space<ul>
<li><a href="stereoisomer_of/">stereoisomer of</a> - a relationship that holds between two forms of the 'same' molecule with identical chemical formulas but different bonds<ul>
<li><a href="diastereoisomers_of/">diastereoisomers of</a> - A stereoisomer non-mirror image non-identical stereoisomers<ul>
<li><a href="epimer_of/">epimer of</a> - A diastereoisomer of relationship in which the isomers differ at only one stereocenter</li>
</ul>
</li>
<li><a href="enantiomer_of/">enantiomer of</a> - a stereoisomer relationship where the two forms are mirror images of one another</li>
</ul>
</li>
<li><a href="structural_isomer_of/">structural isomer of</a> - a relationship that holds between two forms of the 'same' molecule with identical chemical formulas but different bonding patterns<ul>
<li><a href="functional_isomer_of/">functional isomer of</a> - structural isomers that have the same molecular formula (that is, the same number of atoms of the same elements), but the atoms are connected in different ways so that the groupings are dissimilar.</li>
<li><a href="position_isomer_of/">position isomer of</a> - a functional group or other substituent changes position on a parent structure.</li>
<li><a href="skeletal_isomer_of/">skeletal isomer of</a> - components of the (usually carbon) skeleton are distinctly re-ordered to create different structures. Pentane exists as three isomers: n-pentane (often called simply "pentane"), isopentane (2-methylbutane) and neopentane (dimethylpropane).</li>
<li><a href="tautomer_of/">tautomer of</a> - structural isomers (constitutional isomers) of chemical compounds that readily interconvert. Examples: a neutral amino acid and its zwitterion</li>
</ul>
</li>
</ul>
</li>
<li><a href="derivative_of/">derivative of</a><ul>
<li><a href="conjugate_base_of/">conjugate base of</a> - connects the base form of an acid to an acid. Also connects the different charge states of the base. Example: the different charge states of lactate (CHEBI:24996) conjugate base of lactic acid (CHEBI:28358)<ul>
<li><a href="has_acid_form/">has acid form</a> - connects a base to the acid form. E.g. citrate(3-) has acid form citric acid<ul>
<li><a href="acid_base_has_acid_form/">acid base➞has acid form</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="molecule_grouping_class_defined_by_addition_of_a_group_derivative_of/">molecule grouping class defined by addition of a group➞derivative of</a></li>
</ul>
</li>
<li><a href="has_derivative/">has derivative</a><ul>
<li><a href="conjugate_acid_of/">conjugate acid of</a><ul>
<li><a href="acid_form_of/">acid form of</a> - connects an acid to any of the the base anions. E.g. citric acid acid form of citrate(1-),citrate(2-),citrate(3-)<ul>
<li><a href="Bronsted_acid_acid_form_of/">Bronsted acid➞acid form of</a></li>
<li><a href="has_physiological_base/">has physiological base</a> - connects an acid to the base that is stable at physiological pH. E.g. citric acid has physiological base citrate(3-)<ul>
<li><a href="acid_base_conflation_class_has_physiological_base/">acid base conflation class➞has physiological base</a> - The base which defs this conflation class</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><a href="has_element/">has element</a></li>
<li><a href="isobar_of/">isobar of</a> - equal mass number (Z1 + N1 = Z2 + N2)</li>
<li><a href="isotone_of/">isotone of</a> - relationship that holds between two isotones, always holds if the isotones have a different atomic number and the same neutron number.</li>
<li><a href="isotope_of/">isotope of</a> - relationship that holds between two isotopes, always holds if the isotopes have the same atomic number and a different neutron number.</li>
<li><a href="nuclear_isomer_of/">nuclear isomer of</a> - relationship that holds between two nuclear isomers, always holds if the isomers have the same nucleon number but different energy states</li>
</ul>
</li>
<li><a href="classified_by/">classified by</a> - A subtype_of relationship holds between I and C if I is a specific entity of type C. Here I and C denotes instances of OWL classes, where I is a specific class and C is a grouping class</li>
<li><a href="classifies/">classifies</a><ul>
<li><a href="grouping_class_classifies/">grouping class➞classifies</a><ul>
<li><a href="atom_grouping_class_classifies/">atom grouping class➞classifies</a></li>
<li><a href="molecular_component_grouping_class_classifies/">molecular component grouping class➞classifies</a></li>
<li><a href="molecule_grouping_class_classifies/">molecule grouping class➞classifies</a><ul>
<li><a href="chemical_salt_grouping_class_classifies/">chemical salt grouping class➞classifies</a></li>
<li><a href="molecular_derivative_grouping_class_classifies/">molecular derivative grouping class➞classifies</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><a href="has_part/">has part</a> - basic parthood relations<ul>
<li><a href="atom_has_part/">atom➞has part</a></li>
<li><a href="copolymer_has_part/">copolymer➞has part</a></li>
<li><a href="has_atoms/">has atoms</a> - Relation between a molecule and the atoms it contains. Note that this is a shortcut relation: for granular representations go via atom occurrences</li>
<li><a href="has_ionic_component/">has ionic component</a><ul>
<li><a href="has_anionic_component/">has anionic component</a><ul>
<li><a href="chemical_salt_has_anionic_component/">chemical salt➞has anionic component</a></li>
</ul>
</li>
<li><a href="has_cationic_component/">has cationic component</a><ul>
<li><a href="chemical_salt_has_cationic_component/">chemical salt➞has cationic component</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="has_submolecules/">has submolecules</a> - Relation between a molecule and the molecules it is made from.<ul>
<li><a href="has_group/">has group</a> - Relation between a molecule and the groups it contains<ul>
<li><a href="molecule_grouping_class_defined_by_addition_of_a_group_has_group/">molecule grouping class defined by addition of a group➞has group</a></li>
</ul>
</li>
<li><a href="macromolecule_has_submolecules/">macromolecule➞has submolecules</a></li>
</ul>
</li>
<li><a href="material_has_part/">material➞has part</a></li>
<li><a href="molecule_grouping_class_defined_by_components_has_part/">molecule grouping class defined by components➞has part</a></li>
<li><a href="molecule_has_part/">molecule➞has part</a></li>
<li><a href="monomolecular_polymer_polymer_of/">monomolecular polymer➞polymer of</a></li>
<li><a href="polymer_of/">polymer of</a></li>
</ul>
</li>
<li><a href="has_subtype/">has subtype</a></li>
<li><a href="subtype_of/">subtype of</a> - A subtype_of relationship holds between C and P if C is a more specialized form of P. Here C and P denotes instances of OWL classes, where C and P both function as grouping class<ul>
<li><a href="grouping_class_subtype_of/">grouping class➞subtype of</a><ul>
<li><a href="atom_grouping_class_subtype_of/">atom grouping class➞subtype of</a></li>
<li><a href="chemical_grouping_by_charge_subtype_of/">chemical grouping by charge➞subtype of</a> - grouping classes can be arranged in hierarchies arbitrarily deep. Note that subtype_of should only connect two grouping classes</li>
<li><a href="molecular_component_grouping_class_subtype_of/">molecular component grouping class➞subtype of</a></li>
<li><a href="molecule_grouping_class_subtype_of/">molecule grouping class➞subtype of</a> - grouping classes can be arranged in hierarchies arbitrarily deep. Note that subtype_of should only connect two grouping classes</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><a href="chirality_agnostic_form/">chirality agnostic form</a><ul>
<li><a href="racemic_mixture_chirality_agnostic_form/">racemic mixture➞chirality agnostic form</a></li>
</ul>
</li>
<li><a href="classification_relationship/">classification relationship</a><ul>
<li><a href="owl_subclass_of/">owl subclass of</a> - Holds between C and P if C owl:subClassOf P. Note that in this model, most classes are metaclasses (see 'owl class'), where these metaclasses may represent <em>groupings</em> or specific countable elements or species. We provide two different sub-properties of subclass of for two cases: (1) subtype_of, which connects grouping classes (e.g. alkane is_a hydrocarbon; alpha-amino-acid is_a amino-acid), (2) classified_by, which connects a specific element to a grouping class (e.g. methane is_a hydrocarbon; iron is_a metal; serine is_a amino-acid)<ul>
<li><a href="enantiomer_form_of/">enantiomer form of</a> - Example:  R-thalidomide is the enantiomer form of 2-(2,6-dioxopiperidin-3-yl)-1H-isoindole-1,3(2H)-dione</li>
<li><a href="enantiomer_enantiomer_form_of/">enantiomer➞enantiomer form of</a> - Example:  R-thalidomide is the enantiomer form of 2-(2,6-dioxopiperidin-3-yl)-1H-isoindole-1,3(2H)-dione</li>
</ul>
</li>
</ul>
</li>
<li><a href="composed_of/">composed of</a><ul>
<li><a href="proportional_part_composed_of/">proportional part➞composed of</a></li>
</ul>
</li>
<li><a href="coordination_number/">coordination number</a> - The number of atoms bonded to the atom of interest. [https://doi.org/10.1021%2Fed083p791]<ul>
<li><a href="atom_occurrence_coordination_number/">atom occurrence➞coordination number</a></li>
</ul>
</li>
<li><a href="description/">description</a><ul>
<li><a href="reaction_description/">reaction➞description</a></li>
</ul>
</li>
<li><a href="direction/">direction</a> - One of l-&gt;r, r-&gt;l, bidirectional, neutral<ul>
<li><a href="reaction_direction/">reaction➞direction</a></li>
</ul>
</li>
<li><a href="electron_configuration/">electron configuration</a><ul>
<li><a href="chemical_element_electron_configuration/">chemical element➞electron configuration</a></li>
</ul>
</li>
<li><a href="elemental_change/">elemental change</a><ul>
<li><a href="atom_ionic_form_elemental_change/">atom ionic form➞elemental change</a></li>
</ul>
</li>
<li><a href="energy_level/">energy level</a> - e.g. ground, stable</li>
<li><a href="formal_charge/">formal charge</a> - The charge remaining on an atom when all ligands are removed homolytically. [https://doi.org/10.1021%2Fed083p791]<ul>
<li><a href="atom_occurrence_formal_charge/">atom occurrence➞formal charge</a></li>
</ul>
</li>
<li><a href="hard_or_soft/">hard or soft</a><ul>
<li><a href="Bronsted_acid_hard_or_soft/">Bronsted acid➞hard or soft</a></li>
</ul>
</li>
<li><a href="has_atom_occurrences/">has atom occurrences</a><ul>
<li><a href="atomic_bond_has_atom_occurrences/">atomic bond➞has atom occurrences</a></li>
<li><a href="molecule_has_atom_occurrences/">molecule➞has atom occurrences</a></li>
</ul>
</li>
<li><a href="has_bonding_structure/">has bonding structure</a> - The pattern in which the atoms are bonded; e.g. hexagonal lattice, sheets, etc<ul>
<li><a href="allotrope_has_bonding_structure/">allotrope➞has bonding structure</a></li>
</ul>
</li>
<li><a href="has_bonds/">has bonds</a><ul>
<li><a href="molecule_has_bonds/">molecule➞has bonds</a> - The set of all bonds connecting atoms in a molecule</li>
</ul>
</li>
<li><a href="has_charge_state/">has charge state</a><ul>
<li><a href="chemical_grouping_by_charge_has_charge_state/">chemical grouping by charge➞has charge state</a></li>
</ul>
</li>
<li><a href="has_chemical_role/">has chemical role</a></li>
<li><a href="has_left_enantiomer/">has left enantiomer</a><ul>
<li><a href="racemic_mixture_has_left_enantiomer/">racemic mixture➞has left enantiomer</a></li>
</ul>
</li>
<li><a href="has_morphological_category/">has morphological category</a> - e.g. rod, sheet<ul>
<li><a href="nanostructure_has_morphological_category/">nanostructure➞has morphological category</a></li>
</ul>
</li>
<li><a href="has_parent_acid/">has parent acid</a><ul>
<li><a href="ester_has_parent_acid/">ester➞has parent acid</a></li>
</ul>
</li>
<li><a href="has_parent_alcohol/">has parent alcohol</a><ul>
<li><a href="ester_has_parent_alcohol/">ester➞has parent alcohol</a></li>
</ul>
</li>
<li><a href="has_physiologically_stable_form/">has physiologically stable form</a><ul>
<li><a href="chemical_element_has_physiologically_stable_form/">chemical element➞has physiologically stable form</a></li>
</ul>
</li>
<li><a href="has_proportional_parts/">has proportional parts</a> - Relationship between a complex mixture and its parts</li>
<li><a href="has_right_enantiomer/">has right enantiomer</a><ul>
<li><a href="racemic_mixture_has_right_enantiomer/">racemic mixture➞has right enantiomer</a></li>
</ul>
</li>
<li><a href="has_role/">has role</a><ul>
<li><a href="proportional_part_has_role/">proportional part➞has role</a></li>
</ul>
</li>
<li><a href="has_stereocenter/">has stereocenter</a><ul>
<li><a href="chemical_element_has_stereocenter/">chemical element➞has stereocenter</a></li>
</ul>
</li>
<li><a href="in_periodic_table_block/">in periodic table block</a><ul>
<li><a href="chemical_element_in_periodic_table_block/">chemical element➞in periodic table block</a> - block of periodic table unified by the orbitals their valence electrons or vacancies lie in</li>
</ul>
</li>
<li><a href="in_periodic_table_group/">in periodic table group</a><ul>
<li><a href="chemical_element_in_periodic_table_group/">chemical element➞in periodic table group</a> - which number in periodic table column</li>
</ul>
</li>
<li><a href="information_property/">information property</a><ul>
<li><a href="chemical_representation/">chemical representation</a><ul>
<li><a href="AZE_notation_html/">AZE notation html</a> - E.g. <super>12</super>/<sub>6</sub>C</li>
<li><a href="carboydrate_representation/">carboydrate representation</a><ul>
<li><a href="carbbank_representation/">carbbank representation</a> - The carbohydrate structure format used by the Complex Carbohydrate Structure Database (CCSD) originally developed at the University of Georgia. This is a 2D structure where bars and hyphens are used to indicate linkages, and monosaccharides are specified in the following order, separated by hyphens: anomer - configuration symbol - monosaccharide abbreviation.</li>
<li><a href="glycoct_condensed/">glycoct condensed</a> - This is a multiline format used by GlycomeDB, among others, and uses a graph notation to represent glycans, similar to KCF. Monosaccharides are specified more strictly to be able to map the monosaccharides between different representations. Substituents, such as N-acetyl, are specified as separate nodes from their basetypes and are linked together in the edge section. If newlines are removed, this format can be made linear.</li>
<li><a href="glycominds_linearcode_representation/">glycominds linearcode representation</a> - A linear notation proposed by GlycoMinds, Ltd., where monosaccharides are indicated by a one- or two-letter code, linkages are indicated by “a” or “b” for anomers and a single number for the carbon number linkage on the reducing end, and branches are indicated by parentheses. This format is the most compact among existing formats, and several databases use a modified form of this format to accommodate additional codes and varying carbohydrate structures (such as ambiguous structures).</li>
<li><a href="kegg_chemical_function_representation/">kegg chemical function representation</a> - A multiline format used by the KEGG GLYCAN database and represents glycans as graphs. Monosaccharides are represented as nodes, and glycosidic linkages are represented as edges. Each node is given a unique number, and x- and y- coordinates are specified such that the glycan structure can be drawn on a 2D plane. Edge information uses the node numbers to specify anomeric and carbon number information. Monosaccharide names are specified as text, and no particular rules are specified to represent these names.</li>
<li><a href="linucs_representation/">linucs representation</a> - A unique linear notation proposed and used by GLYCOSCIENCES.de, with the reducing end specified on the left. Monosaccharides, written according to CarbBank rules, are surrounded by square brackets and anomers, and stereochemistry is specified and separated by hyphens. Glycosidic linkages are also surrounded by hyphens, and carbon numbers of the linkages are separated by the plus sign. Branches are indicated by curly brackets.</li>
<li><a href="wurcs_representation/">wurcs representation</a></li>
</ul>
</li>
<li><a href="chemical_formula/">chemical formula</a> - A generic grouping for miolecular formulae and empirical formulae<ul>
<li><a href="curly_smiles_string/">curly smiles string</a> - CurlySMILES is a chemical line notation which extends SMILES with annotations for storage, retrieval and modeling of interlinked, coordinated, assembled and adsorbed molecules in supramolecular structures and nanodevices. https://jcheminf.biomedcentral.com/articles/10.1186/1758-2946-3-1</li>
<li><a href="empirical_formula/">empirical formula</a> - A chemical formula where numbers denote relative proportion of atoms. E.g. CH2O for glucose</li>
<li><a href="generalized_empirical_formula/">generalized empirical formula</a> - A variant of an empirical formula where numbers are replaced by generic ns; e.g. CnH2n+2 for alkane</li>
<li><a href="has_rinchi_representation/">has rinchi representation</a> - Proposed Reactions InChI</li>
<li><a href="inchi_string/">inchi string</a></li>
<li><a href="lychi_string/">lychi string</a></li>
<li><a href="molecular_formula/">molecular formula</a> - A chemical formula where numbers denote atom count. E.g. C6H12O6  for glucose</li>
<li><a href="selfies_string/">selfies string</a></li>
<li><a href="smarts_string/">smarts string</a><ul>
<li><a href="reaction_smarts_string/">reaction➞smarts string</a></li>
</ul>
</li>
<li><a href="smiles_string/">smiles string</a> - A string encoding of a molecular graph, no chiral or isotopic information. There are usually a large number of valid SMILES which represent a given structure. For example, CCO, OCC and C(O)C all specify the structure of ethanol.<ul>
<li><a href="canonical_smiles_string/">canonical smiles string</a> - A deterministic SMILES string. Note there is no unique system for canonical smiles strings</li>
<li><a href="extended_smiles_string/">extended smiles string</a></li>
<li><a href="isomeric_smiles_string/">isomeric smiles string</a> - A SMILES string that distinguishes between isomeric forms. E.g. [13C] for carbon-13</li>
</ul>
</li>
</ul>
</li>
<li><a href="has_minchi_representation/">has minchi representation</a> - Proposed new Mixtures InChI</li>
<li><a href="has_rauxinfo_representation/">has rauxinfo representation</a> - Auxhiliary information on top of a RInChi</li>
<li><a href="has_sequence_representation/">has sequence representation</a><ul>
<li><a href="amino_acid_sequence_interval_has_sequence_representation/">amino acid sequence interval➞has sequence representation</a></li>
<li><a href="nucleotide_sequence_interval_has_sequence_representation/">nucleotide sequence interval➞has sequence representation</a> - By convention the sequence string is always the DNA sequence. For RNA intervals, this is the cDNA sequence</li>
</ul>
</li>
<li><a href="hashed_chemical_formula/">hashed chemical formula</a> - A hashed version of a chemical formula<ul>
<li><a href="inchi_key_string/">inchi key string</a> - A hashed representation of an InChi string</li>
<li><a href="rinchi_key_string/">rinchi key string</a> - A hashed representation of a RInChi string<ul>
<li><a href="long_rinchi_key_string/">long rinchi key string</a></li>
<li><a href="short_rinchi_key_string/">short rinchi key string</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><a href="has_mixfile_location/">has mixfile location</a> - Location of a MixFile representing a chemical mixture</li>
<li><a href="has_molfile_location/">has molfile location</a> - Location of a MolFile representing a chemical structure</li>
<li><a href="id/">id</a></li>
<li><a href="inchi_component/">inchi component</a> - A part of an InChI object<ul>
<li><a href="inchi_sublayer/">inchi sublayer</a><ul>
<li><a href="inchi_charge_sublayer/">inchi charge sublayer</a></li>
<li><a href="inchi_fixed_H_layer/">inchi fixed H layer</a></li>
<li><a href="inchi_isotopic_layer/">inchi isotopic layer</a></li>
<li><a href="inchi_proton_sublayer/">inchi proton sublayer</a></li>
<li><a href="inchi_reconnected_layer/">inchi reconnected layer</a></li>
<li><a href="inchi_stereochemical_double_bond_sublayer/">inchi stereochemical double bond sublayer</a></li>
<li><a href="inchi_stereochemical_type_sublayer/">inchi stereochemical type sublayer</a></li>
<li><a href="inchi_sublayer_in_main_layer/">inchi sublayer in main layer</a> - The main layer consists of chemical formula, atom connections, and hydrogen atoms. Example: CH3O/c1-2/h1H3/q-1<ul>
<li><a href="inchi_atom_connections_sublayer/">inchi atom connections sublayer</a> - The atoms in the chemical formula (except for hydrogens) are numbered in sequence; this sublayer describes which atoms are connected by bonds to which other ones.<ul>
<li><a href="polyatomic_entity_inchi_atom_connections_sublayer/">polyatomic entity➞inchi atom connections sublayer</a></li>
</ul>
</li>
<li><a href="inchi_chemical_sublayer/">inchi chemical sublayer</a> - the main chemical formula<ul>
<li><a href="chemical_entity_inchi_chemical_sublayer/">chemical entity➞inchi chemical sublayer</a></li>
</ul>
</li>
<li><a href="inchi_hydrogen_connections_sublayer/">inchi hydrogen connections sublayer</a> - Describes how many hydrogen atoms are connected to each of the other atoms.</li>
</ul>
</li>
<li><a href="inchi_tetrahedral_stereochemical_sublayer/">inchi tetrahedral stereochemical sublayer</a></li>
</ul>
</li>
<li><a href="inchi_version_string/">inchi version string</a><ul>
<li><a href="standard_inchi_object_inchi_version_string/">standard inchi object➞inchi version string</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="name/">name</a> - name of chemical entity. E.g. nickel, carbon-16<ul>
<li><a href="acid_anion_grouping_class_name/">acid anion grouping class➞name</a></li>
<li><a href="chemical_salt_grouping_class_name/">chemical salt grouping class➞name</a><ul>
<li><a href="chemical_salt_by_anion_name/">chemical salt by anion➞name</a></li>
<li><a href="chemical_salt_by_cation_name/">chemical salt by cation➞name</a></li>
</ul>
</li>
<li><a href="molecular_derivative_grouping_class_name/">molecular derivative grouping class➞name</a></li>
</ul>
</li>
<li><a href="symbol/">symbol</a> - short symbol for chemical entity, e.g. K, C-16<ul>
<li><a href="chemical_element_symbol/">chemical element➞symbol</a> - E.g. O, K, Ge</li>
</ul>
</li>
</ul>
</li>
<li><a href="is_balanced/">is balanced</a><ul>
<li><a href="reaction_is_balanced/">reaction➞is balanced</a></li>
</ul>
</li>
<li><a href="is_diastereoselective/">is diastereoselective</a><ul>
<li><a href="reaction_is_diastereoselective/">reaction➞is diastereoselective</a></li>
</ul>
</li>
<li><a href="is_fully_characterized/">is fully characterized</a> - False if includes R-groups<ul>
<li><a href="reaction_is_fully_characterized/">reaction➞is fully characterized</a></li>
</ul>
</li>
<li><a href="is_organic/">is organic</a> - An organic molecule is a molecule composed of organic atoms (at least carbon, hydrogen, and optionally oxygen, phosphorus, nitrogen, sulfur) [SIO]</li>
<li><a href="is_radical/">is radical</a></li>
<li><a href="is_stereo/">is stereo</a><ul>
<li><a href="reaction_is_stereo/">reaction➞is stereo</a></li>
</ul>
</li>
<li><a href="is_substitutent_group_from/">is substitutent group from</a><ul>
<li><a href="functional_group_is_substitutent_group_from/">functional group➞is substitutent group from</a></li>
</ul>
</li>
<li><a href="is_transport/">is transport</a><ul>
<li><a href="reaction_is_transport/">reaction➞is transport</a></li>
</ul>
</li>
<li><a href="left_molecule/">left molecule</a><ul>
<li><a href="molecule_pairwise_similarity_left_molecule/">molecule pairwise similarity➞left molecule</a></li>
</ul>
</li>
<li><a href="left_participants/">left participants</a><ul>
<li><a href="reaction_left_participants/">reaction➞left participants</a></li>
</ul>
</li>
<li><a href="mass/">mass</a><ul>
<li><a href="chemical_element_mass/">chemical element➞mass</a></li>
</ul>
</li>
<li><a href="maximum_percentage/">maximum percentage</a><ul>
<li><a href="proportional_part_maximum_percentage/">proportional part➞maximum percentage</a></li>
</ul>
</li>
<li><a href="melting_point_in_celcius/">melting point in celcius</a><ul>
<li><a href="chemical_element_melting_point_in_celcius/">chemical element➞melting point in celcius</a></li>
</ul>
</li>
<li><a href="minimal_percentage/">minimal percentage</a><ul>
<li><a href="proportional_part_minimal_percentage/">proportional part➞minimal percentage</a></li>
</ul>
</li>
<li><a href="monoisotopic_mass/">monoisotopic mass</a><ul>
<li><a href="chemical_element_monoisotopic_mass/">chemical element➞monoisotopic mass</a></li>
</ul>
</li>
<li><a href="next_in_homologous_series_from/">next in homologous series from</a> - Next in a sequence of compounds with the same functional group and similar chemical properties in which the members of the series can be branched or unbranched. E.g.  ethane (C2H6) next in homologous series from methane (CH4)</li>
<li><a href="occurrence_of/">occurrence of</a><ul>
<li><a href="atom_occurrence_occurrence_of/">atom occurrence➞occurrence of</a></li>
</ul>
</li>
<li><a href="optical_configuration/">optical configuration</a><ul>
<li><a href="enantiomer_optical_configuration/">enantiomer➞optical configuration</a></li>
</ul>
</li>
<li><a href="oxidation_number/">oxidation number</a> - The charge remaining on an atom when all ligands are removed heterolytically in their closed form, with the electrons being transferred to the more electronegative partner; homonuclear bonds do not contribute to the oxidation number. [https://doi.org/10.1021%2Fed083p791]<ul>
<li><a href="atom_occurrence_oxidation_number/">atom occurrence➞oxidation number</a></li>
</ul>
</li>
<li><a href="participant/">participant</a><ul>
<li><a href="reaction_participant_participant/">reaction participant➞participant</a></li>
</ul>
</li>
<li><a href="reaction_center/">reaction center</a><ul>
<li><a href="reaction_reaction_center/">reaction➞reaction center</a></li>
</ul>
</li>
<li><a href="relative_configuration/">relative configuration</a><ul>
<li><a href="enantiomer_relative_configuration/">enantiomer➞relative configuration</a></li>
</ul>
</li>
<li><a href="right_molecule/">right molecule</a><ul>
<li><a href="molecule_pairwise_similarity_right_molecule/">molecule pairwise similarity➞right molecule</a></li>
</ul>
</li>
<li><a href="right_participants/">right participants</a><ul>
<li><a href="reaction_right_participants/">reaction➞right participants</a></li>
</ul>
</li>
<li><a href="score/">score</a><ul>
<li><a href="molecule_pairwise_similarity_score/">molecule pairwise similarity➞score</a><ul>
<li><a href="tanimoto_similarity_score/">tanimoto similarity➞score</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="standard_atomic_weight/">standard atomic weight</a><ul>
<li><a href="chemical_element_standard_atomic_weight/">chemical element➞standard atomic weight</a></li>
</ul>
</li>
<li><a href="stoichiometry/">stoichiometry</a><ul>
<li><a href="reaction_participant_stoichiometry/">reaction participant➞stoichiometry</a></li>
</ul>
</li>
<li><a href="torsional_angle/">torsional angle</a><ul>
<li><a href="atomic_bond_torsional_angle/">atomic bond➞torsional angle</a></li>
</ul>
</li>
<li><a href="type/">type</a><ul>
<li><a href="periodic_table_block_type/">periodic table block➞type</a></li>
</ul>
</li>
<li><a href="valence/">valence</a> - the number of electrons that an atom uses in bonding [https://doi.org/10.1021%2Fed083p791]<ul>
<li><a href="atom_occurrence_valence/">atom occurrence➞valence</a></li>
</ul>
</li>
</ul>
<h3 id="enums">Enums</h3>
<ul>
<li><a href="bond_type_enum/">bond_type_enum</a></li>
<li><a href="hard_or_soft_enum/">hard_or_soft_enum</a></li>
<li><a href="ingredient_role_enum/">ingredient_role_enum</a></li>
<li><a href="nanostructure_morphology_enum/">nanostructure_morphology_enum</a></li>
<li><a href="periodic_table_block_enum/">periodic_table_block_enum</a></li>
</ul>
<h3 id="subsets">Subsets</h3>
<h3 id="types">Types</h3>
<h4 id="built-in">Built in</h4>
<ul>
<li><strong>Bool</strong></li>
<li><strong>Decimal</strong></li>
<li><strong>ElementIdentifier</strong></li>
<li><strong>NCName</strong></li>
<li><strong>NodeIdentifier</strong></li>
<li><strong>URI</strong></li>
<li><strong>URIorCURIE</strong></li>
<li><strong>XSDDate</strong></li>
<li><strong>XSDDateTime</strong></li>
<li><strong>XSDTime</strong></li>
<li><strong>float</strong></li>
<li><strong>int</strong></li>
<li><strong>str</strong></li>
</ul>
<h4 id="defined">Defined</h4>
<ul>
<li><a href="types/DNASequenceString/">DNASequenceString</a>  (<strong>str</strong>) </li>
<li><a href="types/AminoAcidSequenceString/">AminoAcidSequenceString</a>  (<strong>str</strong>) </li>
<li><a href="types/Boolean/">Boolean</a>  (<strong>Bool</strong>)  - A binary (true or false) value</li>
<li><a href="types/ChemicalEncoding/">ChemicalEncoding</a>  (<strong>str</strong>) </li>
<li><a href="types/Count/">Count</a>  (<strong>int</strong>) </li>
<li><a href="types/Date/">Date</a>  (<strong>XSDDate</strong>)  - a date (year, month and day) in an idealized calendar</li>
<li><a href="types/Datetime/">Datetime</a>  (<strong>XSDDateTime</strong>)  - The combination of a date and time</li>
<li><a href="types/Decimal/">Decimal</a>  (<strong>Decimal</strong>)  - A real number with arbitrary precision that conforms to the xsd:decimal specification</li>
<li><a href="types/Double/">Double</a>  (<strong>float</strong>)  - A real number that conforms to the xsd:double specification</li>
<li><a href="types/FileLocation/">FileLocation</a>  (<strong>str</strong>) </li>
<li><a href="types/Float/">Float</a>  (<strong>float</strong>)  - A real number that conforms to the xsd:float specification</li>
<li><a href="types/Integer/">Integer</a>  (<strong>int</strong>)  - An integer</li>
<li><a href="types/Ncname/">Ncname</a>  (<strong>NCName</strong>)  - Prefix part of CURIE</li>
<li><a href="types/Nodeidentifier/">Nodeidentifier</a>  (<strong>NodeIdentifier</strong>)  - A URI, CURIE or BNODE that represents a node in a model.</li>
<li><a href="types/NumberOfYears/">NumberOfYears</a>  (<strong>int</strong>) </li>
<li><a href="types/Objectidentifier/">Objectidentifier</a>  (<strong>ElementIdentifier</strong>)  - A URI or CURIE that represents an object in the model.</li>
<li><a href="types/PeriodicTableGroup/">PeriodicTableGroup</a>  (<strong>int</strong>) </li>
<li><a href="types/SequenceString/">SequenceString</a>  (<strong>str</strong>) </li>
<li><a href="types/String/">String</a>  (<strong>str</strong>)  - A character string</li>
<li><a href="types/Time/">Time</a>  (<strong>XSDTime</strong>)  - A time object represents a (local) time of day, independent of any particular day</li>
<li><a href="types/Uri/">Uri</a>  (<strong>URI</strong>)  - a complete URI</li>
<li><a href="types/Uriorcurie/">Uriorcurie</a>  (<strong>URIorCURIE</strong>)  - a URI or a CURIE</li>
</ul>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="ontology/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Ontology" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Ontology
            </div>
          </div>
        </a>
      
      
        
        <a href="metaclasses/" class="md-footer__link md-footer__link--next" aria-label="Next: Metaclasses" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Metaclasses
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": ".", "features": ["content.tabs.link"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "assets/javascripts/workers/search.22074ed6.min.js"}</script>
    
    
      <script src="assets/javascripts/bundle.960e086b.min.js"></script>
      
    
  </body>
</html>